### ISSUE TITLE
Caused by: java.lang.ClassNotFoundException: Didn't find class "dagger.hilt.internal.GeneratedComponentManagerHolder"

### ISSUE THREAD

**Issue by herriojr** [google/dagger#2697]
I ran into this exception when I was adding ViewBinding to a Fragment. Mind you, this fragment hasn't even been integrated with hilt yet. When I stash my changes on the Fragment and run again, I have no issues. My assumption is that somehow it expects the class in the first Dex file and somehow it was thrown over some limit.

```
dependencies {
    implementation "com.google.dagger:hilt-android:2.35.1"
    kapt "com.google.dagger:hilt-android-compiler:2.35.1"
    androidTestImplementation "com.google.dagger:hilt-android-testing:2.35.1"
    kaptAndroidTest "com.google.dagger:hilt-android-testing:2.35.1"
}
```

```
2021-06-15 14:15:09.423 27259-27259/<suppressed> E/AndroidRuntime: FATAL EXCEPTION: main
    Process: <suppressed>, PID: 27259
    java.lang.RuntimeException: Unable to instantiate application <suppressed>.application.<suppressed>Application: java.lang.ClassNotFoundException: Didn't find class "<suppressed>.application.<suppressed>Application" on path: DexPathList[[zip file "/data/app/~~hjFal36aaFNjCLZ3I4YeGA==/<suppressed>-zJ4QCHWQfUW7vmVzQEVV3g==/base.apk"],nativeLibraryDirectories=[/data/app/~~hjFal36aaFNjCLZ3I4YeGA==/<suppressed>-zJ4QCHWQfUW7vmVzQEVV3g==/lib/x86_64, /data/app/~~hjFal36aaFNjCLZ3I4YeGA==/<suppressed>-zJ4QCHWQfUW7vmVzQEVV3g==/base.apk!/lib/x86_64, /system/lib64, /system_ext/lib64]]
        at android.app.LoadedApk.makeApplication(LoadedApk.java:1244)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6683)
        at android.app.ActivityThread.access$1300(ActivityThread.java:237)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1913)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
     Caused by: java.lang.ClassNotFoundException: Didn't find class "<suppressed>.application.<suppressed>Application" on path: DexPathList[[zip file "/data/app/~~hjFal36aaFNjCLZ3I4YeGA==/<suppressed>-zJ4QCHWQfUW7vmVzQEVV3g==/base.apk"],nativeLibraryDirectories=[/data/app/~~hjFal36aaFNjCLZ3I4YeGA==/<suppressed>-zJ4QCHWQfUW7vmVzQEVV3g==/lib/x86_64, /data/app/~~hjFal36aaFNjCLZ3I4YeGA==/<suppressed>-zJ4QCHWQfUW7vmVzQEVV3g==/base.apk!/lib/x86_64, /system/lib64, /system_ext/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:207)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        at android.app.AppComponentFactory.instantiateApplication(AppComponentFactory.java:76)
        at android.app.Instrumentation.newApplication(Instrumentation.java:1158)
        at android.app.LoadedApk.makeApplication(LoadedApk.java:1236)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6683) 
        at android.app.ActivityThread.access$1300(ActivityThread.java:237) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1913) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:223) 
        at android.app.ActivityThread.main(ActivityThread.java:7656) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947) 
    	Suppressed: java.lang.NoClassDefFoundError: Failed resolution of: Lcom/firstrepublic/banking/application/Hilt_<suppressed>Application;
        at java.lang.VMClassLoader.findLoadedClass(Native Method)
        at java.lang.ClassLoader.findLoadedClass(ClassLoader.java:738)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:363)
        		... 13 more
     Caused by: java.lang.ClassNotFoundException: <suppressed>.application.Hilt_<suppressed>Application
        at java.lang.VMClassLoader.findLoadedClass(Native Method)
        at java.lang.ClassLoader.findLoadedClass(ClassLoader.java:738)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:363)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        		... 16 more
     Caused by: java.lang.NoClassDefFoundError: Failed resolution of: Ldagger/hilt/internal/GeneratedComponentManagerHolder;
        		... 16 more
     Caused by: java.lang.ClassNotFoundException: Didn't find class "dagger.hilt.internal.GeneratedComponentManagerHolder" on path: DexPathList[[zip file "/data/app/~~hjFal36aaFNjCLZ3I4YeGA==/<suppressed>-zJ4QCHWQfUW7vmVzQEVV3g==/base.apk"],nativeLibraryDirectories=[/data/app/~~hjFal36aaFNjCLZ3I4YeGA==/<suppressed>-zJ4QCHWQfUW7vmVzQEVV3g==/lib/x86_64, /data/app/~~hjFal36aaFNjCLZ3I4YeGA==/<suppressed>-zJ4QCHWQfUW7vmVzQEVV3g==/base.apk!/lib/x86_64, /system/lib64, /system_ext/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:207)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        		... 16 more
```

**Comment by bcorso** [comment]
Hmm, it's not clear to me that this is an issue in Hilt or what we could do about it on our side. I would try StackOverflow instead. 

Feel free to reopen if there is something you think we can do in Hilt to fix this.

**Comment by herriojr** [comment]
@bcorso specifically it is unable to find a dagger internal generated class. This feels very reminiscent of some very old Android bugs where there was an expectation that the classes would be in the first dex file. Specifically the only thing I did was refactor a single Fragment to use ViewBinding and make heavy use of, in kotlin, .apply{} for setting items in the view hierarchy. That was it. My guess was that somehow the added anonymous classes somehow moved the dagger.hilt.internal.GeneratedComponentManagerHolder to a location in the dex files that was "inappropriate" for use.

I wish I had kept around the generated apk to dig through the dex files, but unfortunately, I had to undo what I did so I could move forward. No need to reopen at this point, but I'll check back if I find myself in a similar state again.
